File: backend/src/main/java/com/cangchu/product/entity/Sku.java
package com.cangchu.product.entity;

import com.baomidou.mybatisplus.annotation.*;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.fasterxml.jackson.databind.ser.std.ToStringSerializer;
import lombok.Data;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * 商品 SKU（phase-1 A2：公开价三件套 + 上下架）。
 * tenant_id 由 MetaObjectHandler 自动填充；listed 默认 true。
 * spu_id 可空（phase-1 不强制平台 SPU，name+spec 自由录入）。
 */
@Data
@TableName("skus")
public class Sku {

    @TableId(type = IdType.ASSIGN_ID)
    @JsonSerialize(using = ToStringSerializer.class)
    private Long id;

    @JsonSerialize(using = ToStringSerializer.class)
    private Long wholesalerId;

    @JsonSerialize(using = ToStringSerializer.class)
    @TableField(fill = FieldFill.INSERT)
    private Long tenantId;

    /** 平台 SPU（phase-1 可空；P5-D D56 可挂接 ACTIVE 标品） */
    @JsonSerialize(using = ToStringSerializer.class)
    private Long spuId;

    /** 标品名称快照（挂接/合并时整体刷新，22 §2.3；以 OPS 标品为准） */
    private String spuName;

    /** 一级品类快照 */
    private String spuCategoryL1;

    /** 二级品类快照 */
    private String spuCategoryL2;

    private String name;

    /** 规格文本（历史自由文本；P6 批量生成的聚合 SKU 落规格摘要，展示用） */
    private String spec;

    /** 结构化规格键（P6/V42）：模板维度=取值 的有序紧凑 JSON，同 SPU 下唯一；历史/平台挂接为 NULL */
    private String specKey;

    /** 单价（公开价，>0） */
    private BigDecimal unitPrice;

    /** 起批价（>=0） */
    private BigDecimal moqPrice;

    /** 起批量（>=1） */
    private Integer moqQty;

    /** 上下架（默认 true=在售） */
    private Boolean listed;

    /** 主图（phase-1 可空） */
    private String mainImage;

    @TableField(fill = FieldFill.INSERT)
    private LocalDateTime createdAt;

    @TableField(fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime updatedAt;

    private Long createdBy;

    @TableLogic(value = "null", delval = "now()")
    private LocalDateTime deletedAt;
}

File: frontend/packages/api-types/src/sku.ts
/**
 * 商品 SKU 接口 TS 类型（phase-1 D1a 卖家侧上架）
 *
 * 权威来源：后端实现（单一事实源）
 *  - Controller：backend/.../product/controller/SkuController.java
 *  - DTO：SkuCreateDto / SkuUpdateDto
 *  - VO：SkuVo
 *
 * 接口（WA 或该租户 TA 登录态；归属在 service 内以 user_roles 推导）：
 *  - POST /api/v1/tenant/skus?wholesalerId=        创建（wholesalerId 走 query，非 body）
 *  - PUT  /api/v1/tenant/skus/{id}                 修改
 *  - PUT  /api/v1/tenant/skus/{id}/listing?on=     上下架（on=true 上架 / false 下架）
 *  - GET  /api/v1/tenant/skus?wholesalerId=        商户看自己的 SKU（含下架）
 *  - GET  /api/v1/tenant/skus/listed?wholesalerId= 只读：本租户在售（RT 视角，可选）
 *
 * ⚠️ 雪花 ID 字段均为 string（后端 ToStringSerializer）。
 *    价格字段后端为 BigDecimal，JSON 下发为 number/string；前端按 number 处理。
 */

import type { SnowflakeId } from './common'

/** SKU 视图对象（SkuVo） */
export interface Sku {
  id: SnowflakeId
  wholesalerId: SnowflakeId
  tenantId: SnowflakeId
  /** SPU（phase-1 可空） */
  spuId: SnowflakeId | null
  name: string
  spec: string | null
  /** 公开价：单价 */
  unitPrice: number
  /** 公开价：起批价 */
  moqPrice: number | null
  /** 公开价：起批量 */
  moqQty: number | null
  /** 是否在售（上架） */
  listed: boolean
  mainImage: string | null
  createdAt: string
}

/**
 * 创建 SKU 请求（SkuCreateDto，不含 wholesalerId —— 后端走 query 参数）。
 * 后端校验：name @NotBlank ≤128；unitPrice >0；moqPrice ≥0；moqQty ≥1。
 */
export interface CreateSkuRequest {
  name: string
  spec?: string
  unitPrice: number
  moqPrice?: number
  moqQty?: number
  mainImage?: string
}

/** 修改 SKU 请求（SkuUpdateDto，null/缺省表示不改） */
export interface UpdateSkuRequest {
  name?: string
  spec?: string
  unitPrice?: number
  moqPrice?: number
  moqQty?: number
  mainImage?: string
}
